runtime/nvqir/QIRTypes.cpp

```diff
@@ -19,6 +19,42 @@ extern bool initialized;
 extern bool verbose;
 }
 
+namespace {
+// Singleton struct to track allocated arrays
+struct ArrayTracker {
+  // Get the singleton instance of ArrayTracker
+  static ArrayTracker &getInstance() {
+    static ArrayTracker instance;
+    return instance;
+  }
+
+  // Track the given array
+  void track(Array *arr) { allocated_arrays.push_back(arr); }
+  // Untrack the given array (manually delete outside)
+  void untrack(Array *arr) {
+    auto it = std::find(allocated_arrays.begin(), allocated_arrays.end(), arr);
+    if (it != allocated_arrays.end())
+      allocated_arrays.erase(it);
```

Comment:
Should the `else` case here raise an error? It's arguably a very broken program to deallocate arrays that aren't arrays or already dead. Also, it would be cheaper in terms of performance to replace the pointer with a `nullptr` than the resize the vector.

Reply:
I've updated the code to use `nullptr` instead of resizing. I updated the code to emit a warning if an unexpected pointer is provided here. As a clean-up utility, I'm a bit hesitant to make it a hard error.